### PR TITLE
Use Kustomize to build elastic-agent manifests for both managed and standalone more

### LINKED ISSUE #1668: Use Kustomize.io to build elastic-agent manifests for k8s
Similarly to https://github.com/elastic/beats/issues/33514 for Beats, we should make sure that Service Account and RBAC resources are created before DaemonSet and Deployment that use those RBAC permissions otherwise Pods might be in pending states for up to 6 minutes (eg. when running on Kind clusters).

We should replace the [Makefile](https://github.com/elastic/elastic-agent/blob/main/deploy/kubernetes/Makefile) with Kustomize.io

### PR DESCRIPTION
**Pull request by gsantoro**
## What does this PR do?

Use Kustomize to simplify building manifest for elastic-agent both for managed and standalone mode.

## Why is it important?

This PR reduces the amount of boiler plate code needed to build manifests for elastic-agent. Some enhancements:
- we can also install kube-state-metrics just by adding an extra resource `github.com/kubernetes/kube-state-metrics?ref=main` to common/kustomization.yml
- most of the resources (aka roles, role-binding, service accounts, cluster-role, cluster-role binding) are exactly the same for both standalone and managed mode. This reduces massively the amount of duplication.
- some resources (like config maps and extra roles and role bindings) are only necessary in standalone mode. There is no need to duplicates them in both mode
- it makes possible to customise environment variables without changing the manifests . Default values for those environment variables are provided in .env files under dev-tools/kubernetes/base/*/.env (added to git) while custom values are under dev-tools/kubernetes/overlays/*/.env (those should not be under git. that's why they are added to .gitignore). 
- namespace references are only present in kustomization.yml files. It is easier to deploy all resources to a separate namespace
- overlays can be used to customize the docker image tag of the elastic-agent without changing the manifest parts
- Common resources have generic names so that you can use prefixes specific to the mode you want to deploy (eg. `elastic-agent-standalone-` or `elastic-agent-managed`). This enables to deploy standalone and managed agents side by side.
- comments in manifest parts are removed in the final manifest
- yaml code is pretty formatted when the final manifest is generated

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)

## Author's Checklist

- [ ] You can deploy elastic-agent in both standalone and managed modes

## Related issues

- Closes #1668

### CODE REVIEW

File: dev-tools/kubernetes/base/elastic-agent-managed/daemonset.yaml [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+              name: configs
+          securityContext:
+            runAsUser: 0
+          resources:
+            limits:
+              memory: 500Mi
+            requests:
+              cpu: 100m
+              memory: 200Mi
+          volumeMounts:
```
Comment (gizas) [inline dev-tools/kubernetes/base/elastic-agent-managed/daemonset.yaml:67]: I think following part can also be splitted to multilple overalys. 

We want to have needed mounts for observability and extra ones that imposed from security

File: dev-tools/kubernetes/base/elastic-agent-standalone/.env [hunk 2]
```diff
@@ -0,0 +1,3 @@
+ES_USERNAME=elastic
+ES_PASSWORD=changeme
+ES_HOST=https://elasticsearch:9200
```
Comment (gizas) [inline dev-tools/kubernetes/base/elastic-agent-standalone/.env:3]: I would suggest to put namespace as variable here

File: dev-tools/kubernetes/base/elastic-agent-managed/daemonset.yaml (hunk 1 above)
Comment (gsantoro) [reply dev-tools/kubernetes/base/elastic-agent-managed/daemonset.yaml:67]: do you mind elaborate a bit more on the topic. I vaguely recollect some suggestions from the security team but I wouldn't know how to change this part of the manifest

File: dev-tools/kubernetes/base/elastic-agent-standalone/.env (hunk 2 above)
[the remark this answers is not in the document]
Comment (gsantoro) [reply dev-tools/kubernetes/base/elastic-agent-standalone/.env:3]: I don't think so. This env file is only used to create a configmap and I don't think we can use an env variable from a configmap to decide if we want to add a manifest based on that. We could use an overlay with an extra resource to install or not the kube-state-metrics

Comment (gsantoro) [reply dev-tools/kubernetes/base/elastic-agent-standalone/.env:3]: I don't think you can use a variable from an .env file in the kustomization.yml file (which is where I set the namespace for all the resources)

[the remark this answers is not in the document]
Comment (gizas) [reply dev-tools/kubernetes/base/elastic-agent-standalone/.env:3]: But then we will need to edit the kustomize.yaml.

Maybe we can think that we can provide diffrent kustomzation files ?
eg kustomize-ksmenabled,  kustomize-ksmdisabled?

[the remark this answers is not in the document]
Comment (gizas) [reply dev-tools/kubernetes/base/elastic-agent-standalone/.env:3]: Also a tool worth looking: 
https://carvel.dev/ytt/

So we would like to provide some basic YAML templating and then kustomize installation

**Review by gizas — APPROVED** [review]

File: dev-tools/kubernetes/base/elastic-agent-managed/daemonset.yaml [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+                  fieldPath: spec.nodeName
+            - name: POD_NAME
+              valueFrom:
+                fieldRef:
+                  fieldPath: metadata.name
+          envFrom:
+          - configMapRef:
+              # Fleet Server URL to enroll the Elastic Agent into
+              # FLEET_URL can be found in Kibana, go to Management > Fleet > Settings
+              # - name: FLEET_URL
```
Comment (ChrsMark) [inline dev-tools/kubernetes/base/elastic-agent-managed/daemonset.yaml:41]: Just out of curiosity, why all these settings are commented out?

Comment (gsantoro) [reply dev-tools/kubernetes/base/elastic-agent-managed/daemonset.yaml:41]: they are mostly left there for documentation. We should address them (one way or another) in the next update about on this topic.

Comment (ChrsMark) [reply dev-tools/kubernetes/base/elastic-agent-managed/daemonset.yaml:41]: But when those vars should be uncommented?

Comment (gsantoro) [reply dev-tools/kubernetes/base/elastic-agent-managed/daemonset.yaml:41]: I have created this [PR](https://github.com/elastic/elastic-agent/pull/2160) to fix this

**Pull request merged**